### PR TITLE
New attribute authority storage

### PR DESCRIPTION
**Pull request by AleDore**

### CODE REVIEW

File: openapi/index.yaml [hunk 1]
```diff
@@ -11,45 +11,153 @@ host: api.cd.italia.it
...
+      responses:
+        "200":
+          description: Inserted referent.
+        "404":
+          description: Not found.
+        "500":
+          description: Service unavailable.
+          schema:
+            $ref: "#/definitions/ProblemJson"
+    delete:
```
Comment (AleDore) [inline openapi/index.yaml:127]: I suppose this operation must be under an endpoint that accepts referent fiscalCode in path

File: src/server.ts [hunk 2]
```diff
@@ -27,46 +39,124 @@ const blobServiceClient = BlobServiceClient.fromConnectionString(
...
+        request,
+        reply,
+        pathParamsMiddleware(KeyOrganizationFiscalCode),
+        requiredBodyMiddleware(ReferentFiscalCode)
+      ),
+  },
+  referentHandler.insertReferentHandler()
+);
+
+server.delete<{ Body: ReferentFiscalCode }>(
```
Comment (AleDore) [inline src/server.ts]: ditto from api specs

File: src/services/companyService.ts [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
-  blobName: NonEmptyString
+  fiscalCode: FiscalCode
 ): TE.TaskEither<Error, O.Option<Companies>> =>
   pipe(
-    getBlobData(blobServiceClient, containerName, blobName, UsersCompanies),
-    TE.map(
-      A.findFirstMap(elem =>
-        elem.fiscalCode === fiscalCode ? O.some(elem.companies) : O.none
+    TE.tryCatch(
+      async () =>
```
Comment (AleDore) [inline src/services/companyService.ts:16]: do we need async modifier?

File: src/services/organizationService.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+export const upsertOrganization = (
+  organizationWithReferents: OrganizationWithReferents
+): TE.TaskEither<Error, OrganizationWithReferents> =>
+  pipe(
+    TE.of(
+      organizationWithReferents.keyOrganizationFiscalCode !==
+        organizationWithReferents.organizationFiscalCode
+    ),
+    TE.chain((shouldUpdatePrimaryKey) =>
+      TE.tryCatch(async () => {
```
Comment (AleDore) [inline src/services/organizationService.ts]: ditto for async

File: src/services/organizationService.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                },
+                type: QueryTypes.UPDATE,
+              }
+            )
+          : void 0;
+      }, E.toError)
+    ),
+    TE.chain(() =>
+      TE.tryCatch(
+        async () =>
```
Comment (AleDore) [inline src/services/organizationService.ts]: ditto

File: src/services/organizationService.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            await Promise.all(
+              organizationWithReferents.referents.map((r) =>
+                Referent.upsert({ fiscalCode: r })
+              )
+            )
+          ).map(([r, _]) => r);
+          await organization.addReferents(referents, {
+            ignoreDuplicates: true,
+          });
+        }, E.toError),
```
Comment (AleDore) [inline src/services/organizationService.ts]: why not refactor this implementation into a chain based one?

**Review by AleDore — COMMENTED** [review]
In general try avoiding to explicit async await into chain definitions.

File: src/services/organizationService.ts (hunk 6 above)
Comment (michaeldisaro) [reply src/services/organizationService.ts]: in https://github.com/pagopa/ade-aa-ms-mock/pull/16/commits/9bd4e8454fe620851cd808de302de858df4fb62c

File: src/services/organizationService.ts (hunk 5 above)
Comment (michaeldisaro) [reply src/services/organizationService.ts]: in https://github.com/pagopa/ade-aa-ms-mock/pull/16/commits/9bd4e8454fe620851cd808de302de858df4fb62c

File: src/services/organizationService.ts (hunk 4 above)
Comment (michaeldisaro) [reply src/services/organizationService.ts]: in https://github.com/pagopa/ade-aa-ms-mock/pull/16/commits/9bd4e8454fe620851cd808de302de858df4fb62c

File: src/services/companyService.ts (hunk 3 above)
Comment (michaeldisaro) [reply src/services/companyService.ts:16]: in https://github.com/pagopa/ade-aa-ms-mock/pull/16/commits/9bd4e8454fe620851cd808de302de858df4fb62c

File: src/server.ts (hunk 2 above)
Comment (michaeldisaro) [reply src/server.ts]: in https://github.com/pagopa/ade-aa-ms-mock/pull/16/commits/9bd4e8454fe620851cd808de302de858df4fb62c

File: openapi/index.yaml (hunk 1 above)
Comment (michaeldisaro) [reply openapi/index.yaml:127]: in https://github.com/pagopa/ade-aa-ms-mock/pull/16/commits/9bd4e8454fe620851cd808de302de858df4fb62c

**Pull request merged**